pkg/codegen/hcl2/model/binder_expression.go

```diff
@@ -370,15 +379,16 @@ func (b *expressionBinder) bindForExpression(syntax *hclsyntax.ForExpr) (Express
 			syntax.CondExpr != nil)
 	}
 	expr := &ForExpression{
-		Syntax:        syntax,
-		Tokens:        tokens,
-		KeyVariable:   keyVariable,
-		ValueVariable: valueVariable,
-		Collection:    collection,
-		Key:           key,
-		Value:         value,
-		Condition:     condition,
-		Group:         syntax.Group,
+		Syntax:                       syntax,
+		Tokens:                       tokens,
+		KeyVariable:                  keyVariable,
+		ValueVariable:                valueVariable,
+		Collection:                   collection,
+		Key:                          key,
+		Value:                        value,
+		Condition:                    condition,
+		Group:                        syntax.Group,
+		StrictCollectionTypechecking: !b.options.skipRangeTypechecking,
```

Comment:
Bit sad these have to end up on the expression tree, but looks to be very awkward to avoid that :(

Reply:
Yeah :/